File: backend/schemas.py
# backend/schemas.py
# Removed PlantingGroup schemas and updated Planting schemas.
from pydantic import BaseModel, ConfigDict, create_model
from typing import Optional, List, Type
from datetime import date, datetime
from models import PlantingStatus, PlantingMethod, TaskStatus, HarvestMethod
import enum


# --- Helper function for creating update schemas ---
def make_optional(model: Type[BaseModel]) -> Type[BaseModel]:
    """Creates a new Pydantic model with all fields from the original model made optional."""
    fields = model.model_fields
    optional_fields = {name: (Optional[field.annotation], None) for name, field in fields.items()}
    return create_model(f'{model.__name__}Update', **optional_fields)

class ImportResult(BaseModel):
    message: str
    imported_count: int = 0
    updated_count: int = 0
    skipped_count: int = 0
    errors: List[str] = []

# --- Plant Schemas ---
class PlantBase(BaseModel):
    plant_name: str
    variety_name: Optional[str] = None
    scientific_name: Optional[str] = None
    plant_family: Optional[str] = None
    plant_type: Optional[str] = None
    growth_habit: Optional[str] = None
    origin_heirloom_status: Optional[str] = None
    organic: Optional[bool] = False
    seed_company_source: Optional[str] = None
    year_acquired: Optional[int] = None
    seed_size: Optional[str] = None
    seed_longevity_storage_life: Optional[str] = None
    germination_temperature_min: Optional[int] = None
    germination_temperature_max: Optional[int] = None
    germination_temperature_ideal: Optional[int] = None
    germination_time_days: Optional[str] = None
    light_requirement_for_germination: Optional[str] = None
    stratification_required: Optional[bool] = False
    scarification_required: Optional[bool] = False
    sowing_depth: Optional[str] = None
    spacing_in_row: Optional[str] = None
    spacing_low: Optional[int] = None
    spacing_high: Optional[int] = None
    direct_seedable: Optional[bool] = False
    transplantable: Optional[bool] = False
    days_to_transplant_low: Optional[int] = None
    days_to_transplant_high: Optional[int] = None
    time_to_maturity: Optional[str] = None
    mature_plant_height: Optional[str] = None
    mature_plant_spread_width: Optional[str] = None
    sunlight_requirement: Optional[str] = None
    water_needs: Optional[str] = None
    fertilizer_needs: Optional[str] = None
    pest_resistance: Optional[str] = None
    disease_resistance: Optional[str] = None
    cold_hardiness_frost_tolerance: Optional[str] = None
    heat_tolerance: Optional[str] = None
    drought_tolerance: Optional[str] = None
    bolting_tendency: Optional[str] = None
    support_required: Optional[str] = None
    pruning_required: Optional[str] = None
    harvest_window_low: Optional[int] = None
    harvest_window_high: Optional[int] = None
    typical_yield: Optional[str] = None
    yield_units: Optional[str] = None
    storage_life_post_harvest: Optional[str] = None
    requires_pollinator: Optional[bool] = False
    notes_observations: Optional[str] = None
    url: Optional[str] = None
    weekly_yield: Optional[str] = None

class PlantCreate(PlantBase):
    pass

PlantUpdate = make_optional(PlantBase)

class Plant(PlantBase):
    id: int
    model_config = ConfigDict(from_attributes=True)

# --- Task Group Schemas ---
class TaskGroupBase(BaseModel):
    pass

class TaskGroupCreate(TaskGroupBase):
    pass

class TaskGroup(TaskGroupBase):
    id: int
    tasks: List["Task"] = []
    model_config = ConfigDict(from_attributes=True)

# --- Recurring Task Schemas ---
class RecurringTaskBase(BaseModel):
    name: str
    description: Optional[str] = None
    recurrence_rule: str  # RRULE string
    recurrence_end_date: Optional[date] = None

class RecurringTaskCreate(RecurringTaskBase):
    garden_plan_id: int
    planting_id: Optional[int] = None
    start_date: date
    exdates: Optional[List[date]] = []

class RecurringTaskUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    recurrence_rule: Optional[str] = None
    recurrence_end_date: Optional[date] = None
    exdates: Optional[List[date]] = None

# --- Task Schemas ---
class TaskBase(BaseModel):
    name: str
    description: Optional[str] = None
    due_date: Optional[date] = None
    status: TaskStatus = TaskStatus.PENDING

class TaskCreate(TaskBase):
    garden_plan_id: int
    planting_id: Optional[int] = None
    task_group_id: Optional[int] = None
    recurring_task_id: Optional[int] = None

class TaskUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    due_date: Optional[date] = None
    status: Optional[TaskStatus] = None
    task_group_id: Optional[int] = None

class Task(TaskBase):
    id: int
    garden_plan_id: int
    planting_id: Optional[int] = None
    task_group_id: Optional[int] = None
    recurring_task_id: Optional[int] = None
    model_config = ConfigDict(from_attributes=True)

class RecurringTask(RecurringTaskBase):
    id: int
    garden_plan_id: int
    planting_id: Optional[int] = None
    tasks: List[Task] = []
    model_config = ConfigDict(from_attributes=True)

# --- Planting Schemas ---
class Planting(PlantBase):
    id: int
    library_plant_id: int
    quantity: int
    status: PlantingStatus
    planned_sow_date: Optional[date] = None
    planned_transplant_date: Optional[date] = None
    planned_harvest_start_date: Optional[date] = None
    planned_harvest_end_date: Optional[date] = None
    planned_second_harvest_date: Optional[date] = None
    time_to_maturity_override: Optional[int] = None
    planting_method: Optional[PlantingMethod] = None
    harvest_method: Optional[HarvestMethod] = None
    tasks: List[Task] = []
    recurring_tasks: List[RecurringTask] = []
    model_config = ConfigDict(from_attributes=True)

class PlantingCreate(BaseModel):
    library_plant_id: int
    quantity: int = 1
    status: PlantingStatus = PlantingStatus.PLANNED
    planting_method: Optional[PlantingMethod] = None
    harvest_method: Optional[HarvestMethod] = None
    time_to_maturity_override: Optional[int] = None
    planned_sow_date: Optional[date] = None
    planned_transplant_date: Optional[date] = None
    planned_harvest_start_date: Optional[date] = None
    planned_harvest_end_date: Optional[date] = None
    planned_second_harvest_date: Optional[date] = None

PlantingUpdateSchema = make_optional(Planting)

class PlantingDetail(Planting):
    model_config = ConfigDict(from_attributes=True)

# --- Action Schemas ---
class LogActionType(str, enum.Enum):
    SOW_SEEDS = "Sow Seeds"
    TRANSPLANT = "Transplant"
    FIRST_HARVEST = "First Harvest"

class LogAction(BaseModel):
    action_type: LogActionType
    action_date: date
    quantity: int = 1

# --- Garden Plan Schemas ---
class GardenPlanBase(BaseModel):
    name: str
    description: Optional[str] = None

class GardenPlanCreate(GardenPlanBase):
    pass

GardenPlanUpdate = make_optional(GardenPlanBase)

class GardenPlan(GardenPlanBase):
    id: int
    created_date: date
    last_accessed_date: datetime
    plantings: List[Planting] = []
    tasks: List[Task] = []
    recurring_tasks: List[RecurringTask] = []
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/schemas.ts
// frontend/src/schemas.ts
import { z } from 'zod';

// Enums
export const PlantingStatusSchema = z.enum([
  "Planned",
  "Started",
  "Transplanted",
  "Direct Sown",
  "Growing",
  "Harvesting",
  "Done",
]);

export const PlantingMethodSchema = z.enum([
  "Seed Starting",
  "Direct Seeding",
  "Seedling",
]);

export const HarvestMethodSchema = z.enum([
  "Single Harvest",
  "Cut and Come Again",
  "Staggered",
  "Continuous",
]);

export const LogActionTypeSchema = z.enum([
  "Sow Seeds",
  "Transplant",
  "First Harvest",
]);

export const TaskStatusSchema = z.enum([
  "Pending",
  "In Progress",
  "Completed",
]);

// Schemas
export const PlantSchema = z.object({
  id: z.number(),
  plant_name: z.string(),
  variety_name: z.string().nullable().optional(),
  scientific_name: z.string().nullable().optional(),
  plant_family: z.string().nullable().optional(),
  plant_type: z.string().nullable().optional(),
  growth_habit: z.string().nullable().optional(),
  origin_heirloom_status: z.string().nullable().optional(),
  organic: z.boolean().nullable().optional(),
  seed_company_source: z.string().nullable().optional(),
  year_acquired: z.number().nullable().optional(),
  seed_size: z.string().nullable().optional(),
  seed_longevity_storage_life: z.string().nullable().optional(),
  germination_temperature_min: z.number().nullable().optional(),
  germination_temperature_max: z.number().nullable().optional(),
  germination_temperature_ideal: z.number().nullable().optional(),
  germination_time_days: z.string().nullable().optional(),
  light_requirement_for_germination: z.string().nullable().optional(),
  stratification_required: z.boolean().nullable().optional(),
  scarification_required: z.boolean().nullable().optional(),
  sowing_depth: z.string().nullable().optional(),
  spacing_in_row: z.string().nullable().optional(),
  spacing_low: z.number().nullable().optional(),
  spacing_high: z.number().nullable().optional(),
  direct_seedable: z.boolean().nullable().optional(),
  transplantable: z.boolean().nullable().optional(),
  days_to_transplant_low: z.number().nullable().optional(),
  days_to_transplant_high: z.number().nullable().optional(),
  time_to_maturity: z.string().nullable().optional(),
  mature_plant_height: z.string().nullable().optional(),
  mature_plant_spread_width: z.string().nullable().optional(),
  sunlight_requirement: z.string().nullable().optional(),
  water_needs: z.string().nullable().optional(),
  fertilizer_needs: z.string().nullable().optional(),
  pest_resistance: z.string().nullable().optional(),
  disease_resistance: z.string().nullable().optional(),
  cold_hardiness_frost_tolerance: z.string().nullable().optional(),
  heat_tolerance: z.string().nullable().optional(),
  drought_tolerance: z.string().nullable().optional(),
  bolting_tendency: z.string().nullable().optional(),
  support_required: z.string().nullable().optional(),
  pruning_required: z.string().nullable().optional(),
  harvest_window_low: z.number().nullable().optional(),
  harvest_window_high: z.number().nullable().optional(),
  typical_yield: z.string().nullable().optional(),
  yield_units: z.string().nullable().optional(),
  storage_life_post_harvest: z.string().nullable().optional(),
  requires_pollinator: z.boolean().nullable().optional(),
  notes_observations: z.string().nullable().optional(),
  url: z.string().nullable().optional(),
  weekly_yield: z.string().nullable().optional(),
});

export const TaskSchema = z.object({
  id: z.number(),
  garden_plan_id: z.number(),
  planting_id: z.number().nullable().optional(),
  task_group_id: z.number().nullable().optional(),
  recurring_task_id: z.number().nullable().optional(),
  name: z.string(),
  description: z.string().nullable().optional(),
  due_date: z.string().nullable().optional(),
  status: TaskStatusSchema,
});

export const RecurringTaskSchema = z.object({
  id: z.number(),
  garden_plan_id: z.number(),
  planting_id: z.number().optional(),
  name: z.string(),
  description: z.string().optional(),
  recurrence_rule: z.string(),
  recurrence_end_date: z.string().optional(),
  exdates: z.array(z.string()).optional(),
  tasks: z.array(TaskSchema),
});

export const PlantingSchema = PlantSchema.extend({
  quantity: z.number(),
  status: PlantingStatusSchema,
  library_plant_id: z.number(),
  planned_sow_date: z.string().nullable().optional(),
  planned_transplant_date: z.string().nullable().optional(),
  planned_harvest_start_date: z.string().nullable().optional(),
  planned_harvest_end_date: z.string().nullable().optional(),
  planned_second_harvest_date: z.string().nullable().optional(),
  time_to_maturity_override: z.number().nullable().optional(),
  planting_method: PlantingMethodSchema.nullable().optional(),
  harvest_method: HarvestMethodSchema.nullable().optional(),
  tasks: z.array(TaskSchema),
  recurring_tasks: z.array(RecurringTaskSchema),
});

export const PlantingCreatePayloadSchema = z.object({
  library_plant_id: z.number(),
  quantity: z.number(),
  status: PlantingStatusSchema,
  planting_method: PlantingMethodSchema.nullable().optional(),
  harvest_method: HarvestMethodSchema.nullable().optional(),
  time_to_maturity_override: z.number().nullable().optional(),
  planned_sow_date: z.string().optional(),
  planned_transplant_date: z.string().optional(),
  planned_harvest_start_date: z.string().optional(),
  planned_harvest_end_date: z.string().optional(),
  planned_second_harvest_date: z.string().optional(),
});

export const GardenPlanSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable().optional(),
  created_date: z.string(),
  last_accessed_date: z.string(),
  plantings: z.array(PlantingSchema),
  tasks: z.array(TaskSchema),
  recurring_tasks: z.array(RecurringTaskSchema),
});

export const LoginCredentialsSchema = z.object({
  username: z.string(),
  password: z.string(),
});

export const RegistrationDataSchema = z.object({
  username: z.string(),
  email: z.string().email(),
  password: z.string(),
});

export const UserSchema = z.object({
  id: z.number(),
  username: z.string(),
  email: z.string().email(),
});

// Create inferred types from schemas
export type Plant = z.infer<typeof PlantSchema>;
export type PlantingStatus = z.infer<typeof PlantingStatusSchema>;
export type PlantingMethod = z.infer<typeof PlantingMethodSchema>;
export type HarvestMethod = z.infer<typeof HarvestMethodSchema>;
export type LogActionType = z.infer<typeof LogActionTypeSchema>;
export type TaskStatus = z.infer<typeof TaskStatusSchema>;
export type Task = z.infer<typeof TaskSchema>;
export type RecurringTask = z.infer<typeof RecurringTaskSchema>;
export type Planting = z.infer<typeof PlantingSchema>;
export type PlantingCreatePayload = z.infer<typeof PlantingCreatePayloadSchema>;
export type GardenPlan = z.infer<typeof GardenPlanSchema>;
export type LoginCredentials = z.infer<typeof LoginCredentialsSchema>;
export type RegistrationData = z.infer<typeof RegistrationDataSchema>;
export type User = z.infer<typeof UserSchema>;

// Schema for the form in AddToPlanModal
export const PlantingFormSchema = z.object({
  quantity: z.number().min(1, 'Quantity must be at least 1'),
  plantingMethod: PlantingMethodSchema,
  harvestMethod: HarvestMethodSchema,
  sowDate: z.string().optional(),
  transplantDate: z.string().optional(),
  harvestDate: z.string().optional(),
  harvestEndDate: z.string().optional(),
  secondHarvestDate: z.string().optional(),
  timeToMaturity: z.string().regex(/^\d*$/, "Must be a number").optional(),
  daysToTransplant: z.string().regex(/^\d*$/, "Must be a number").optional(),
}).refine(data => {
    if (data.sowDate && data.transplantDate) {
        return new Date(data.transplantDate) >= new Date(data.sowDate);
    }
    return true;
}, {
    message: "Transplant date must be after sow date.",
    path: ["transplantDate"],
}).refine(data => {
    if (data.transplantDate && data.harvestDate) {
        return new Date(data.harvestDate) >= new Date(data.transplantDate);
    }
    return true;
}, {
    message: "Harvest date must be after transplant date.",
    path: ["harvestDate"],
}).refine(data => {
    if (data.harvestDate && data.harvestEndDate) {
        return new Date(data.harvestEndDate) >= new Date(data.harvestDate);
    }
    return true;
}, {
    message: "Harvest end date must be after harvest start date.",
    path: ["harvestEndDate"],
}).refine(data => {
    if (data.harvestDate && data.secondHarvestDate) {
        return new Date(data.secondHarvestDate) >= new Date(data.harvestDate);
    }
    return true;
}, {
    message: "Second harvest date must be after first harvest date.",
    path: ["secondHarvestDate"],
});

export type PlantingFormData = z.infer<typeof PlantingFormSchema>;

// Schema for the RecurrenceEditor component state
export const RecurrenceOptionsSchema = z.object({
  freq: z.number(),
  interval: z.number().min(1),
  byday: z.any().nullable(),
  bymonthday: z.number().nullable(),
  bysetpos: z.number().nullable(),
  bymonth: z.number().nullable(),
  count: z.number().nullable(),
  until: z.date().nullable(),
  endType: z.enum(['never', 'date', 'count']),
  dailyOption: z.enum(['everyday', 'weekdays']),
  monthlyOption: z.enum(['day_of_month', 'day_of_week']),
}).refine(data => {
    // Add complex validation logic here as needed.
    // For example, ensure byday is set for weekly recurrences.
    if (data.freq === 2 && (!data.byday || data.byday.length === 0)) { // RRule.WEEKLY = 2
        return false;
    }
    return true;
}, {
    message: "For weekly recurrences, at least one day must be selected.",
    path: ["byday"],
});

export type RecurrenceOptions = z.infer<typeof RecurrenceOptionsSchema>;

// Schema for CSV export
export const CsvExportSchema = z.array(z.object({
    library_plant_id: z.number(),
    plant_name: z.string(),
    variety_name: z.string().nullable().transform(v => v || ''),
    quantity: z.number(),
    status: PlantingStatusSchema,
    planting_method: PlantingMethodSchema.nullable().transform(v => v || ''),
    planned_sow_date: z.string().nullable().transform(v => v || ''),
    planned_transplant_date: z.string().nullable().transform(v => v || ''),
    planned_harvest_start_date: z.string().nullable().transform(v => v || ''),
    time_to_maturity_override: z.number().nullable().transform(v => v || ''),
}));

// Schema for parsing days which can be a number, a string, or a range string '10-14'
export const DaysSchema = z.preprocess((val) => {
    if (typeof val === 'string') {
        if (val.trim() === '') return null;
        if (val.includes('-')) {
            const parts = val.split('-').map(d => parseInt(d.trim(), 10));
            if (parts.length === 2 && !parts.some(isNaN)) {
                return Math.round((parts[0] + parts[1]) / 2);
            }
            return null;
        }
        const num = parseInt(val, 10);
        return isNaN(num) ? null : num;
    }
    if (typeof val === 'number') {
        return val;
    }
    return null;
}, z.number().nullable());